### ISSUE TITLE
Regexp highlighter

### ISSUE THREAD

**Issue by disarmer** [zsh-users/zsh-syntax-highlighting#380]
Hi,
I regularly use regexp highlighter, so may be it will be useful for others. It's basically like pattern highlighter, but allow to set more advanced patterns:
https://gist.github.com/disarmer/0d0a2721a9515a51b749ff7bba7b5c4c

With this you can use PCRE regexp for patterns, for example:
`ZSH_HIGHLIGHT_REGEXP+=('\bsudo\b' fg=123,bold)`
This will highlight "sudo" only if it's surrounded by word boundary, i.e. "sudo cmd", but not "sudoedit"

**Comment by danielshahaf** [comment]
Thanks for the patch, @disarmer.

Review:

- The hardcoded `999` is inappropriate; would be better to use `-1` there, or `${buf:$MEND}`.

- I'm concerned about the use of recursion.  zsh has a recursion depth limit; highlighting something like space might run into it.  Maybe switch to an explicit loop?

- The initial call to `_loop` ought to pass `0` for $3.

Moreover, if this is to be merged, documentation (in markdown) would be needed too.

$PREBUFFER is ignored.  (It's non-empty a multiline construct was entered at the prompt, e.g., after typing `{` and `<Enter>` at the prompt.)  Is that intentional?

Bottom line: I think this would be a useful addition; would you be able to package it as a pull request?

Thanks!

**Comment by disarmer** [comment]
Great, i'll fix these and make a pull request in near time.
Hardcoded 999 used as recursion limit, because zsh has 1000 depth limit.
About the use of recursion: i tested corner case: 
`ZSH_HIGHLIGHT_REGEXP+=('-' fg=123,bold)`
I flooded input line with a lot of '-' and zsh got a segfault. Then i lowered 999 and tested several times and found that zsh segfault when this boundary greater 640. I don't know, it's look like bug in zsh, may be because too many escape sequences.
But anyway, it's corner-case, in normal situations user should use '-+' regexp (i.e matching one or more minus sign).
Now if i leave 640 as recursion limit and write comment about this number, it will be ok?

**Comment by danielshahaf** [comment]
> Great, i'll fix these and make a pull request in near time.

I look forward to that.

> Hardcoded 999 used as recursion limit, because zsh has 1000 depth limit.
> About the use of recursion: i tested corner case:
> ZSH_HIGHLIGHT_REGEXP+=('-' fg=123,bold)
> I flooded input line with a lot of '-' and zsh got a segfault. Then i lowered 999 and tested several times and found that zsh segfault when this boundary greater 640. I don't know, it's look like bug in zsh, may be because too many escape sequences.

I expect the segfault is because your zsh was compiled with the `--disable-max-function-depth` configure option.  What's the output of `% f() { f }; f` (at the prompt) in your shell?  Mine gives "f:1: maximum nested function level reached".

> But anyway, it's corner-case, in normal situations user should use '-+' regexp (i.e matching one or more minus sign).

This doesn't change anything; it just changes the problem from "how many minuses" to "how many runs of minuses".

> Now if i leave 640 as recursion limit and write comment about this number, it will be ok?

Sorry, but no.  The value '640' will change from user to user and from zsh release to zsh release; and it'd be better to avoid the issue entirely by rewriting the code iteratively rather than recursively.

Feel free to join us on IRC (#zsh-syntax-highlighting on freenode).

Cheers,

**Comment by disarmer** [comment]
> I expect the segfault is because your zsh was compiled with the --disable-max-function-depth 

No, i just use standard zsh package from archlinux ( here configure options: https://git.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/zsh )
```
% f() { f }; f
f: maximum nested function level reached
```
This not showing depth limit for me, so i modified your code a bit:
```
% f() { let X=X+1;echo $X;f }; f
...
999
1000
f: maximum nested function level reached
```
> This doesn't change anything; it just changes the problem from "how many minuses" to "how many runs of minuses".

But it's change: with '-+' it's match whole line of minuses in one "zsh_highlight_regexp_highlighter_loop" call, then recursive call does nothing (because it's get empty $buf) and return

Ok, i'll try to rewrite to iterative function, but not yet sure how

**Comment by danielshahaf** [comment]
> But it's change: with '-+' it's match whole line of minuses in one "zsh_highlight_regexp_highlighter_loop" call, then recursive call does nothing (because it's get empty $buf) and return

We're talking about different cases.  I was thinking of the case where $BUFFER has many non-consecutive matches, for example, matching when the regexp is `/[ ]/` (without the slashes; they're just delimiters) and $BUFFER contains a shell command.

> Ok, i'll try to rewrite to iterative function, but not yet sure how

To clarify, I mean to rewrite the `_loop` function in terms of a `for` or `while` loop.  Let us know if you need a pointer.

**Comment by disarmer** [comment]
I rewrite to iterative cycle and created a pull request, please take a look:
https://github.com/zsh-users/zsh-syntax-highlighting/pull/381

**Comment by danielshahaf** [comment]
Looks good, thanks!  I'll write a detailed review later.  I'd like to merge #356 next week and then release a beta; it'd be great if we could merge #381 by then, too.

**Comment by danielshahaf** [comment]
Update: haven't forgotten about this.  I can't merge it now because I'd like to keep master in a releaseable state, so as soon as upstream releases 5.3 I can merge #356 and release 0.6.0 here; but merging this branch would require a soak period before a subsequent release.  I'd have merged this branch now if I knew the upstream release was still a few weeks away, but it might not be that far.

Adding the "upstream" label because of this scheduling dependency.  Apologies for the delay.

**Comment by disarmer** [comment]
Great, thank you

**Comment by danielshahaf** [comment]
We're going to release a quick 0.6.0 so this will be deferred to 0.7.0.  Apologies, I should have merged this sooner.

**Comment by phy1729** [comment]
Merged in 4e6f600.
